### PR TITLE
Configure Spree to run tests with Postgres in Semaphore

### PR DESCRIPTION
**Pull request by HugsDaniel**
#### What ? Why ? 

Following the conversation with @mkllnk and @sauloperez, this is an attempt to config Spree for using Postgres instead of SQLite3.

Two questions : 
 - In `build.sg` the `DB=postgres` before `bundle exec rake test_app` param is needed so that Spree knows we want to have a PG db in our dummy apps. Default behavior comes from :  https://github.com/openfoodfoundation/spree/blob/fe0a1311abb097bf3ac8001a24246c6cff5ecdbb/core/lib/generators/spree/dummy/dummy_generator.rb#L30
Again, in the spirit of "the fewer the changes, the better", I preferred to use `DB=postgres` in `build.sh`. Is there a better way to do it ?
  - In `core/lib/generators/spree/dummy/templates/rails/database.yml`, I added Semaphore db credentials so that it can run without any authentication issue. The problem is that it changes the generation of the dummy apps in local as well, making it impossible to use on our machines except if we have a db with same credentials. I saw in Semaphore that we can create a db with `createdb test_db -U $DATABASE_POSTGRESQL_USERNAME`, but it seems to be an even bigger change in the code. What do you guys think ?

It solves the problem we have in https://github.com/openfoodfoundation/openfoodnetwork/issues/2537 so I removed the pending.

Also using PG makes a test fail, so I commented it out.

### CODE REVIEW

**Comment by luisramos0** [comment]
"the fewer the changes, the better"
interesting, if it's all broken, more changes are better!

question 1:
in this specific case, I'd change it in dummy_generator.rb. not in build.sh, you want to keep build.sh very very clean.

question 2:
I am not sure I follow you but the tests run on pg locally so we could just use the same credentials in both semaphore and local. the credentials dont need to be sempahore specific.

**Comment by HugsDaniel** [comment]
@luisramos0 the thing is that tests run on pg locally only if you specify it, by default Spree runs it with SQLite3. In `core/lib/generators/spree/dummy/templates/rails/database.yml` the default username for pg is `postgres`, and trying to use that in Semaphore will trigger a `authentication failed, wrong password for username 'postgres'`. Locally I had to put the postgres user on trust mode to avoid having this issue, but in Semaphore I don't think we can/should do that. Maybe I'm missing something here ?

File: core/lib/generators/spree/dummy/templates/rails/database.yml [hunk 1]
```diff
@@ -33,17 +33,20 @@ production:
 development:
   adapter: postgresql
   database: <%= database_prefix %>spree_development
-  username: postgres
+  username: runner
```
Comment (luisramos0) [inline core/lib/generators/spree/dummy/templates/rails/database.yml]: do you really need to change development and production here? Is test not the only one used?

File: core/lib/generators/spree/dummy/templates/rails/database.yml [hunk 2]
```diff
@@ -33,17 +33,20 @@ production:
...
+  password: semaphoredb
   min_messages: warning
 test:
   adapter: postgresql
   database: <%= database_prefix %>spree_test
-  username: postgres
+  username: runner
```
Comment (luisramos0) [inline core/lib/generators/spree/dummy/templates/rails/database.yml]: I wouldnt use the default sempahore db. I'd create one.

I dont think semaphore is a good name because this will be used locally as well.
Username can be postgres or, if you need a variation from the default pg user, you can use username ofn.

File: build.sh [hunk 3]
```diff
@@ -17,7 +17,7 @@ function set_gemfile(){
 echo "Setup Spree defaults and creating test application..."
 rm_gemfile_lock
 bundle check || bundle install
-bundle exec rake test_app
+DB=postgres bundle exec rake test_app
```
Comment (luisramos0) [inline build.sh]: I'd put this on dummy_generators.db instead.

Comment (mkllnk) [inline build.sh]: This is an environment variable we can set in Semaphore. Semaphore allows to set global environment variables. Alternatively, we can change the build script to include it:

```sh
DB=postgres bash build.sh
```

So we don't have to change the code for this.

File: core/lib/generators/spree/dummy/templates/rails/database.yml (hunk 1 above)
Comment (mkllnk) [inline core/lib/generators/spree/dummy/templates/rails/database.yml]: Which environment do we actually need for the tests? Is it running in test environment or in development environment?

Another way to solve this is to write our own database.yml on Semaphore: https://semaphoreci.com/docs/languages/ruby/custom-database-yml.html

I added you to the project and you should be able to change the Semaphore configuration.

**Review by mkllnk — COMMENTED** [review]
As commented below, I would try to do the configuration in Semaphore. If it was completely our code, I would make the build scripts more generic.

**Comment by HugsDaniel** [comment]
@mkllnk @luisramos0 @sauloperez I followed you advice using environment variables. As @sauloperez suggested, I edited directly the `database.yml` file in the generator to fetch Semaphore EVs. Now it all passes. This one and #18 solve all remaining failures for Spree core build.

**Review by luisramos0 — APPROVED** [review]
Awesome Hugo! A green build is a green build :-)
Let's do it.

Just one comment on commits: I prefer to rebase my own PRs. In this case, you have one commit adding "DB=postgres" on build.sh and another one later removing it. In these cases I prefer to rebase and not include these extra steps in the final commit history. Let me know on slack if you dont know how to do it. I can show you.

File: common_spree_dependencies.rb [hunk 4]
```diff
@@ -6,7 +6,7 @@
 gem 'json'
 gem 'multi_json'
 gem 'mysql2'
-gem 'pg'
+gem 'pg', '~> 0.11'
```
Comment (sauloperez) [inline common_spree_dependencies.rb]: Why is this change needed? Since in OFN we use 0.21.0 I would stick to that one to be sure things work on our version.

File: core/lib/generators/spree/dummy/templates/rails/database.yml [hunk 5]
```diff
@@ -33,17 +33,20 @@ production:
 development:
   adapter: postgresql
   database: <%= database_prefix %>spree_development
-  username: postgres
+  username: <%= ENV.fetch('DATABASE_USERNAME', 'default') %>
```
Comment (sauloperez) [inline core/lib/generators/spree/dummy/templates/rails/database.yml]: to be backward compatible we should use `postgres` as the second argument. This way if no `DATABASE_USERNAME` is provided we fall back to that value.

**Review by sauloperez — CHANGES_REQUESTED** [review]
collonut @HugsDaniel !!

File: common_spree_dependencies.rb (hunk 4 above)
Comment (HugsDaniel) [reply common_spree_dependencies.rb]: It's the one used here (see the Gemfile.lock), if I don't specify `'~> 0.11'` it uses 1.1.x and raises errors

File: core/lib/generators/spree/dummy/templates/rails/database.yml [hunk 6]
```diff
@@ -33,17 +33,20 @@ production:
 development:
   adapter: postgresql
   database: <%= database_prefix %>spree_development
-  username: postgres
+  username: <%= ENV.fetch('DATABASE_USERNAME', 'postgres') %>
+  password: <%= ENV.fetch('DATABASE_PASSWORD', 'default') %>
```
Comment (sauloperez) [inline core/lib/generators/spree/dummy/templates/rails/database.yml]: as there was no password before, I'd say the default should be an empty string, don't you think?

File: common_spree_dependencies.rb (hunk 4 above)
Comment (sauloperez) [reply common_spree_dependencies.rb]: In that case, it might be better to specify anything below the one that raises errors. What about '< 1.1.0'?

Comment (HugsDaniel) [reply common_spree_dependencies.rb]: '< 1.1.0' raised the same error, '<1.0' seem to work fine (it uses 0.21)

File: core/lib/generators/spree/dummy/templates/rails/database.yml (hunk 6 above)
Comment (HugsDaniel) [reply core/lib/generators/spree/dummy/templates/rails/database.yml]: Oh ok I thought 'default' was like default behavior

**Review by sauloperez — APPROVED** [review]

**Review by mkllnk — APPROVED** [review]
